x-pack/plugins/alerting/server/task_runner/rule_loader.test.ts

```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { encryptedSavedObjectsMock } from '@kbn/encrypted-saved-objects-plugin/server/mocks';
+import { KibanaRequest } from '@kbn/core/server';
+import { schema } from '@kbn/config-schema';
+import type { PublicMethodsOf } from '@kbn/utility-types';
+
+import { getDecryptedAttributes, getFakeKibanaRequest, loadRule } from './rule_loader';
+import { TaskRunnerContext } from './task_runner_factory';
+import { ruleTypeRegistryMock } from '../rule_type_registry.mock';
+import { rulesClientMock } from '../rules_client.mock';
+import { RulesClient } from '../rules_client';
+import { Rule } from '../types';
+import { MONITORING_HISTORY_LIMIT, RuleExecutionStatusErrorReasons } from '../../common';
+import { getReasonFromError } from '../lib/error_with_reason';
+import { alertingEventLoggerMock } from '../lib/alerting_event_logger/alerting_event_logger.mock';
+
+// create mocks
+const rulesClient = rulesClientMock.create();
+const ruleTypeRegistry = ruleTypeRegistryMock.create();
+const alertingEventLogger = alertingEventLoggerMock.create();
+const encryptedSavedObjects = encryptedSavedObjectsMock.createClient();
+const mockBasePathService = { set: jest.fn() };
+
+// assign default parameters/data
+const apiKey = 'rule-apikey';
+const ruleId = 'rule-id-1';
+const enabled = true;
+const spaceId = 'rule-spaceId';
+const ruleName = 'rule-name';
+const consumer = 'rule-consumer';
+const ruleTypeId = 'rule-type-id';
+const ruleParams = { paramA: 42 };
+
+describe('rule_loader', () => {
+  let context: TaskRunnerContext;
+  let contextMock: ReturnType<typeof getTaskRunnerContext>;
+
+  const paramValidator = schema.object({
+    paramA: schema.number(),
+  });
+
+  const DefaultLoadRuleParams = {
+    paramValidator,
+    ruleId,
+    spaceId,
+    ruleTypeRegistry,
+    alertingEventLogger,
+  };
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+    encryptedSavedObjects.getDecryptedAsInternalUser.mockImplementation(
+      mockGetDecrypted({
+        apiKey,
+        enabled,
+        consumer,
+      })
+    );
+    contextMock = getTaskRunnerContext(ruleParams, MONITORING_HISTORY_LIMIT);
+    context = contextMock as unknown as TaskRunnerContext;
+  });
+
+  afterEach(() => {
+    jest.restoreAllMocks();
+  });
+
+  describe('loadRule()', () => {
```

Comment:
Do you think we could remove some of the tests inside `task_runner.test.ts` that test for errors when loading the rule since we have these tests here? I think as long as we have test coverage for what happens in the event of these various failures and then one test inside the task runner test for what happens when `loadRule` throws an error, we should be covered right?

Reply:
I'm not real comfortable with that, though it would be a nice goal.  The error tests I saw were also checking some other side effects of the error processing, so we'd be missing those tests.

Were there some specific tests you were thinking of?  Maybe I missed some obvious ones we could remove ...

Reply:
Just quickly looking at the tests, I was thinking:
* `validates params before running the rule type`
* `uses API key when provided`
* `doesn't use API key when not provided`
* `recovers gracefully when the Alert Task Runner throws an exception when fetching the encrypted attributes`
* `recovers gracefully when the Alert Task Runner throws an exception when license is higher than supported`
* `recovers gracefully when the Alert Task Runner throws an exception when getting internal Services`
* `recovers gracefully when the Alert Task Runner throws an exception when fetching attributes`
* `successfully bails on execution if the rule is disabled`

These all seem to test things inside the rule loading? But if you're not comfortable with it, we can save it for a later time.

Reply:
Looking at `validates params before running the rule type`, it's testing the result of `taskRunner.run()`, so we'd lose the test of that.  I'm not sure how critical that is, and we'd have to hunt down if we are testing all these in another way.  Some of the other tests are checking the usageCounter and eventLog - again, not sure.

My current thought is that once we've boiled this module down a bit more, we'll probably see some streamlining we can do, including with the tests.  So feels like not the best time to be doing this.

I can open an issue to track, we could add it to the task runner meta issue ...

Reply:
That sounds good!